### PR TITLE
Fix iOS platform view not deallocated

### PR DESCRIPTION
**Pull request by zhongwuzw**
Fixes https://github.com/flutter/flutter/issues/54937

There has a retain cycle like below:
SemanticsObject->platformViewSemanticsContainer->platformViewSemanticsContainer.accessibilityElements->SemanticsObject

Another two PRs for memory leaks:
https://github.com/flutter/engine/pull/17954
https://github.com/flutter/engine/pull/18107

### CODE REVIEW

File: shell/platform/darwin/ios/framework/Source/accessibility_bridge.mm [hunk 1]
```diff
@@ -242,6 +246,9 @@ static bool DidFlagChange(const flutter::SemanticsNode& oldNode,
 }
 
 void AccessibilityBridge::clearState() {
+  for (SemanticsObject* object in objects_.get().allValues) {
+    object.platformViewSemanticsContainer = nil;
```
Comment (cyanglaz) [inline shell/platform/darwin/ios/framework/Source/accessibility_bridge.mm:250]: We can add this in the `dealloc` of `SemanticsObject`

Comment (zhongwuzw) [reply shell/platform/darwin/ios/framework/Source/accessibility_bridge.mm:250]: Emm, actually we did it in `dealloc` of `SemanticsObject`, https://github.com/flutter/engine/blob/fb208b486ec354dc7324e4228d9c311208174b18/shell/platform/darwin/ios/framework/Source/SemanticsObject.mm#L198, but cause of retain cycle, dealloc not executed.

**Comment by gaaclarke** [comment]
Please add a test.  Previously we haven't been able to easily write tests for this code but now we are able to, see also https://github.com/flutter/engine/blob/master/shell/platform/darwin/ios/framework/Source/SemanticsObjectTest.mm

I'll add a new test file for you in another PR.

**Comment by gaaclarke** [comment]
@zhongwuzw Apologies for the delay.  I landed that some unit tests for that class so it should be easy to write tests =)

**Comment by zhongwuzw** [comment]
@gaaclarke 👍 ,I added test, please review.

File: shell/platform/darwin/ios/framework/Source/accessibility_bridge_test.mm [hunk 2]
```diff
@@ -5,11 +5,61 @@
...
+@end
+
+@interface MockFlutterPlatformFactory : NSObject <FlutterPlatformViewFactory>
+@end
+
+@implementation MockFlutterPlatformFactory
+- (NSObject<FlutterPlatformView>*)createWithFrame:(CGRect)frame
+                                   viewIdentifier:(int64_t)viewId
+                                        arguments:(id _Nullable)args {
+  MockFlutterPlatformView* platformView = [[MockFlutterPlatformView new] autorelease];
```
Comment (gaaclarke) [inline shell/platform/darwin/ios/framework/Source/accessibility_bridge_test.mm:58]: Your example is using autorelease (which generally is proper) but other plugins are not doing this: https://github.com/flutter/plugins/blob/531ff2bc80afb3181218a1ff46e195303348ae0c/packages/google_maps_flutter/google_maps_flutter/ios/Classes/GoogleMapController.m#L37

File: shell/platform/darwin/ios/framework/Source/accessibility_bridge_test.mm [hunk 3]
```diff
@@ -5,11 +5,61 @@
...
+  if (self = [super init]) {
+    MockPlatformView* view = [[MockPlatformView new] autorelease];
+    mockPlatformView = view;
+    self.view = view;
+  }
+  return self;
+}
+
+- (void)dealloc {
+  self.view = nil;
```
[the remark this answers is not in the document]
Comment (gaaclarke) [reply shell/platform/darwin/ios/framework/Source/accessibility_bridge_test.mm:45]: done

File: shell/platform/darwin/ios/framework/Source/accessibility_bridge_test.mm [hunk 4]
```diff
@@ -5,11 +5,61 @@
...
 #import "flutter/shell/platform/darwin/common/framework/Headers/FlutterMacros.h"
+#import "flutter/shell/platform/darwin/ios/framework/Headers/FlutterPlatformViews.h"
+#import "flutter/shell/platform/darwin/ios/framework/Source/FlutterPlatformViews_Internal.h"
 #import "flutter/shell/platform/darwin/ios/framework/Source/accessibility_bridge.h"
 #import "flutter/shell/platform/darwin/ios/platform_view_ios.h"
 #import "third_party/ocmock/Source/OCMock/OCMock.h"
 
 FLUTTER_ASSERT_NOT_ARC
+@class MockPlatformView;
+MockPlatformView* mockPlatformView = nil;
```
[the remark this answers is not in the document]
Comment (gaaclarke) [reply shell/platform/darwin/ios/framework/Source/accessibility_bridge_test.mm]: done

File: shell/platform/darwin/ios/framework/Source/accessibility_bridge_test.mm [hunk 5]
```diff
@@ -131,4 +181,55 @@ - (void)testUpdateSemanticsOneNode {
...
+        FlutterPlatformViewGestureRecognizersBlockingPolicyEager);
+    FlutterResult result = ^(id result) {
+    };
+    flutterPlatformViewsController->OnMethodCall(
+        [FlutterMethodCall
+            methodCallWithMethodName:@"create"
+                           arguments:@{@"id" : @2, @"viewType" : @"MockFlutterPlatformView"}],
+        result);
+
+    __block auto bridge = std::make_unique<flutter::AccessibilityBridge>(
```
[the remark this answers is not in the document]
Comment (gaaclarke) [reply shell/platform/darwin/ios/framework/Source/accessibility_bridge_test.mm]: done

File: shell/platform/darwin/ios/framework/Source/accessibility_bridge_test.mm [hunk 6]
```diff
@@ -131,4 +181,55 @@ - (void)testUpdateSemanticsOneNode {
...
+                                 /*ui=*/thread_task_runner,
+                                 /*io=*/thread_task_runner);
+    auto platform_view = std::make_unique<flutter::PlatformViewIOS>(
+        /*delegate=*/mock_delegate,
+        /*rendering_api=*/flutter::IOSRenderingAPI::kSoftware,
+        /*task_runners=*/runners);
+    id mockFlutterView = OCMClassMock([FlutterView class]);
+    std::string label = "some label";
+
+    flutter::FlutterPlatformViewsController* flutterPlatformViewsController =
```
[the remark this answers is not in the document]
Comment (gaaclarke) [reply shell/platform/darwin/ios/framework/Source/accessibility_bridge_test.mm]: done

File: shell/platform/darwin/ios/framework/Source/accessibility_bridge_test.mm (hunk 2 above)
Comment (gaaclarke) [reply shell/platform/darwin/ios/framework/Source/accessibility_bridge_test.mm]: done

File: shell/platform/darwin/ios/framework/Source/accessibility_bridge_test.mm [hunk 7]
```diff
@@ -131,4 +181,55 @@ - (void)testUpdateSemanticsOneNode {
...
+    semantics_node.id = 2;
+    semantics_node.platformViewId = 2;
+    semantics_node.label = label;
+    nodes[kRootNodeId] = semantics_node;
+    flutter::CustomAccessibilityActionUpdates actions;
+    bridge->UpdateSemantics(/*nodes=*/nodes, /*actions=*/actions);
+    flutterPlatformViewsController->Reset();
+    delete flutterPlatformViewsController;
+  }
+  XCTAssertNil(mockPlatformView);
```
[the remark this answers is not in the document]
Comment (gaaclarke) [reply shell/platform/darwin/ios/framework/Source/accessibility_bridge_test.mm]: done

**Comment by gaaclarke** [comment]
@zhongwuzw Ok, the bug is fixed if you want to give it a look.  The problem with memory leaks is that sometimes people can start depending on them.  Let me double check this still works with the webview plugin.

**Comment by gaaclarke** [comment]
@zhongwuzw Sorry, what I tried seemed correct but broke the plugins.  I reverted it.  Your unit test and thus your patch is still wrong because it relies on autoreleasing the result of `createWithFrame:viewIdentifier:arguments:` which isn't how it's suppose to work.

File: shell/platform/darwin/ios/framework/Source/accessibility_bridge_test.mm (hunk 2 above)
Comment (zhongwuzw) [reply shell/platform/darwin/ios/framework/Source/accessibility_bridge_test.mm:59]: @gaaclarke Emm, I think they are the same thing, they also return autorelease object actually by ARC.

**Comment by zhongwuzw** [comment]
> @zhongwuzw Sorry, what I tried seemed correct but broke the plugins. I reverted it. Your unit test and thus your patch is still wrong because it relies on autoreleasing the result of `createWithFrame:viewIdentifier:arguments:` which isn't how it's suppose to work.

`createWithFrame:viewIdentifier:arguments:` actually always return autorelease object, please refer to https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/MemoryMgmt/Articles/mmRules.html.

File: shell/platform/darwin/ios/framework/Source/accessibility_bridge_test.mm (hunk 2 above)
Comment (gaaclarke) [reply shell/platform/darwin/ios/framework/Source/accessibility_bridge_test.mm:59]: You're right.  Sorry, half of our codebase is in ARC half is not.  I thought I was looking at non-arc code.

**Comment by gaaclarke** [comment]
@zhongwuzw thanks for pointing out my mistake with the ARC code.  I switched the code back to making `FlutterPlatformViewSemanticsContainer.semanticsObject` weak.  I got stuck yesterday because of the ARC issue.

This way is preferred because it adheres to the contract better.  The way it was originally posted the property was strong, but relied in outside manipulation to work.  That was fragile because people had to remember to update that property whenever the `object_` instance variable is modified.  This way that maintenance doesn't have to happen.

I've tested the webview plugin and it works and your test still passes.

**Pull request merged**